### ISSUE TITLE
Guide for using Slanted Triangular Scheduler

### ISSUE THREAD

**Issue by mateuszpieniak** [allenai/allennlp#3670]
**Question**

I don't fully understand the implementation of Slanted Triangular Scheduler - https://github.com/allenai/allennlp/blob/master/allennlp/training/learning_rate_schedulers/slanted_triangular.py. Could you explain what is the effect of `num_epochs` and `num_steps_per_epoch` argument?

Assume my training takes 20 epoch and 100 iterations / steps each. Now let's use Slanted Triangular Scheduler. What happens if I use it with `num_epochs = 2` and `num_steps_per_epoch = 20`? Is it working only for the first 2 epochs and returns 5 triangles for each epoch?

Thanks in advance!

**Comment by DeNeutoy** [comment]
num epochs is the total number of epochs for your training, `num_steps_per_epoch` is the number of grad updates you do in a epoch, e.g dataset_size/batch_size. Slanted triangular learning rate is only one large triangle for your whole training, not multiple triangle schedules.

**Comment by cemilcengiz** [comment]
> num epochs is the total number of epochs for your training, `num_steps_per_epoch` is the number of grad updates you do in a epoch, e.g dataset_size/batch_size. Slanted triangular learning rate is only one large triangle for your whole training, not multiple triangle schedules.

Is this the same for multi-gpu case or should we divide it by the number of gpus i.e. num_steps_per_epoch=dataset_size/(batch_size*num_gpus) ?

**Comment by mateuszpieniak** [comment]
@DeNeutoy Could you relate to my example? What happens if the `num_epochs` and `num_steps_per_epoch` from STLR don't match the `num_epochs` and `batch_size` of my training dataset? 

Is there then a one big triangle only for a part of my training dataset and flat learning rate for the rest?

**Comment by DeNeutoy** [comment]
@cemilcengiz in previous versions the batch size was divided by the number of GPUs internally (i.e batch size was the total, and per gpu would have batch_size/num_gpus batch_size), so it would have been correct. However on master yes, you would need to change this.

@mateuszpieniak Yes, it would just be flat after that I think. You can monitor this by using `should_log_learning_rate` parameter to the trainer, which then allows you to see how the learning rate changes in tensorboard.

**Comment by mateuszpieniak** [comment]
@DeNeutoy Thanks for the tip. I verified how it works and yes, it gives a flat learning rate afterwards. In addition, there are some cases worth commenting on. 

1) A triangle always spans for `learning_rate_scheduler.num_epochs`. If there are 97 "true" iterations in one epoch of training and `learning_rate_scheduler.num_epochs = 2`, then the triangle will span for 194 iterations regardless of `learning_rate_scheduler.num_steps_per_epoch`. 

1a) What happens if I specify `learning_rate_scheduler.num_steps_per_epoch = 10` iterations? Then it's not a triangle anymore, but the span is still the same. See the example below.

![Screenshot from 2020-01-23 11-31-07](https://user-images.githubusercontent.com/31375424/72977057-f4a3f100-3dd3-11ea-86a6-a1971bae3e31.png)

Take in mind that the triangle doesn't end after exactly 10th iteration (is it a bug?), but around ~14th iteration.

1b) What happens if I specify `learning_rate_scheduler.num_steps_per_epoch = 999999` iterations? Then it "delayed" triangle for some reason (is it a bug?). See the example below.

![Screenshot from 2020-01-23 11-42-39](https://user-images.githubusercontent.com/31375424/72977775-7d6f5c80-3dd5-11ea-928c-04e9013960a4.png)

1c) What happens if I specify `learning_rate_scheduler.num_steps_per_epoch = 97` iterations? Then it works as expected.

![Screenshot from 2020-01-23 11-52-33](https://user-images.githubusercontent.com/31375424/72978850-72b5c700-3dd7-11ea-9087-11566623b4f2.png)

1d) What happens if I specify `learning_rate_scheduler.num_epochs` >> `trainer.num_epochs`? Then the triangle still spans for `learning_rate_scheduler.num_epochs`, hence we end up with weird shape.
